chat-plugins/uno.js

```diff
@@ -0,0 +1,547 @@
+				this.eliminate(this.currentPlayer);
+			}, this.maxTime * 1000);
+			break;
+		}
+		if (initialize) this.onNextPlayer();
+	}
+
+	onSelectColour(user, colour) {
+		if (!["Red", "Blue", "Green", "Yellow"].includes(colour) || user.userid !== this.currentPlayer || this.state !== "colour") return false;
+		this.topCard.colour = colour; // manually change the top card's colour
+		this.sendToRoom(`The colour has been changed to ${colour}.`);
+		clearTimeout(this.timer);
+		this.state = "play";
+
+		if (this.isPlusFour) {
+			this.isPlusFour = false;
+			this.onNextPlayer(); // handle the skipping here.
+		}
+
+		this.nextTurn();
+	}
+
+	onDrawCard(user, count, selfDraw) {
+		if (!(user.userid in this.players)) return false;
+		let drawnCards = this.drawCard(count);
+
+		let player = this.players[user.userid];
+		player.hand.push(...drawnCards);
+		player.sendDisplay();
+		player.sendRoom(`|raw|You have drawn the following card${drawnCards.length > 1 ? "s" : ""}: ${drawnCards.map(card => `<span style="color: ${card.colour}">${card.name}</span>`).join(", ")}.`);
+		if (selfDraw) player.cardLock = drawnCards[0].name;
+	}
+
+	drawCard(count) {
+		count = parseInt(count);
+		if (!count || count < 1) count = 1;
+		let drawnCards = [];
+
+		for (let i = 0; i < count; i++) {
+			if (!this.deck.length) {
+				this.deck = this.discards.length ? Tools.shuffle(this.discards) : Tools.shuffle(createDeck()); // shuffle the cards back into the deck, or if there are no discards, add another deck into the game.
+			}
+			drawnCards.push(this.deck.pop());
+		}
+		return drawnCards;
+	}
+
+	onUno(user, unoId) {
+		// uno id makes spamming /uno uno impossible
+		if (this.unoId !== unoId || user.userid !== this.awaitUno) return false;
+		this.sendToRoom(`|raw|<strong>UNO!</strong> ${user.name} is down to their last card!`);
+		delete this.awaitUno;
+		delete this.unoId;
+	}
+
+	onSendHand(user) {
+		if (!(user.userid in this.players) || this.state === "signups") return false;
+
+		this.players[user.userid].sendDisplay();
+	}
+
+	onWin(player) {
+		this.sendToRoom(`|raw|<div class="broadcast-green">Congratulations to ${player.name} for winning the game of UNO!</div>`);
+		this.destroy();
+	}
+
+	destroy() {
+		clearTimeout(this.timer);
+		this.sendToRoom(`|uhtmlchange|uno-${this.room.gameNumber}|<div class="infobox">The game of UNO has ended.</div>`);
+
+		// deallocate games for each player.
+		for (let i in this.players) {
+			this.players[i].destroy();
+		}
+		delete this.room.game;
+	}
+}
+
+class UNOgamePlayer extends Rooms.RoomGamePlayer {
+	constructor(user, game) {
+		super(user, game);
+		this.hand = [];
+	}
+
+	canPlayWildFour() {
+		let colour = this.game.topCard.colour;
+
+		if (this.hand.some(c => c.colour === colour)) return false;
+		return true;
+	}
+
+	hasCard(cardName) {
+		let matched = this.hand.filter(c => c.name === cardName);
+		if (!matched.length) return false;
+		return matched[0]; // returns the first instance of the matched card
```

Comment:
Just ``this.hand.find(c => c.name === cardName)``

(Note that this alternative returns ``undefined`` if there is no match).